libraries/ui/src/ui/types/ContextAwareProfile.cpp

```diff
@@ -20,20 +21,60 @@
 
 static const QString RESTRICTED_FLAG_PROPERTY = "RestrictFileAccess";
 
-ContextAwareProfile::ContextAwareProfile(QQmlContext* context) :
-    ContextAwareProfileParent(context), _context(context) { 
-    assert(context);
+QMutex RestrictedContextMonitor::gl_monitorMapProtect;
+RestrictedContextMonitor::TMonitorMap RestrictedContextMonitor::gl_monitorMap;
+
+RestrictedContextMonitor::~RestrictedContextMonitor() {
+    QMutexLocker locker(&gl_monitorMapProtect);
+    TMonitorMap::iterator lookup = gl_monitorMap.find(_context);
+    if (lookup != gl_monitorMap.end()) {
+        gl_monitorMap.erase(lookup);
+    }
+}
+
+RestrictedContextMonitor::TSharedPointer RestrictedContextMonitor::getMonitor(QQmlContext* context, bool createIfMissing) {
+    TSharedPointer monitor;
+
+    QMutexLocker locker(&gl_monitorMapProtect);
+    TMonitorMap::const_iterator lookup = gl_monitorMap.find(context);
+    if (lookup != gl_monitorMap.end()) {
+        monitor = lookup.value().lock();
+        assert(monitor);
+    } else if(createIfMissing) {
+        monitor = TSharedPointer::create(context);
+        monitor->_selfPointer = monitor;
+        gl_monitorMap.insert(context, monitor);
+    }
+    return monitor;
 }
 
+ContextAwareProfile::ContextAwareProfile(QQmlContext* context) : ContextAwareProfileParent(context), _context(context) {
+    assert(context);
+
+    _monitor = RestrictedContextMonitor::getMonitor(context, true);
+    assert(_monitor);
+    connect(_monitor.get(), &RestrictedContextMonitor::onIsRestrictedChanged, this, &ContextAwareProfile::onIsRestrictedChanged);
+    if (_monitor->_isUninitialized) {
+        _monitor->_isRestricted = isRestrictedGetProperty();
+        _monitor->_isUninitialized = false;
+    }
+    _isRestricted.store(_monitor->_isRestricted ? 1 : 0);
+}
 
 void ContextAwareProfile::restrictContext(QQmlContext* context, bool restrict) {
+    RestrictedContextMonitor::TSharedPointer monitor = RestrictedContextMonitor::getMonitor(context, false);
+
     context->setContextProperty(RESTRICTED_FLAG_PROPERTY, restrict);
+    if (monitor && monitor->_isRestricted != restrict) {
+        monitor->_isRestricted = restrict;
+        monitor->onIsRestrictedChanged(restrict);
```

Comment:
Ok, I understand that the core of the fix is removing the blocking call from ContextAwareProfile::isRestricted(), but I'm still not sure what the purpose of the monitor is.  Couldn't we just directly set ContextAwareProfile::_isRestricted here?

Reply:
Jeez it's been a while since I've looked at this code.  I think the key point of the monitor is this line:

connect(_monitor.get(), &RestrictedContextMonitor::onIsRestrictedChanged, this, &ContextAwareProfile::onIsRestrictedChanged);

I'm thinking I've got one monitor object for each QQmlContext, which serves as something for ContextAwareProfile to subscribe to changes to.  I'm assuming with this approach there may be many more ContextAwareProfile's than QQmlContext's.

The RestrictedContextMonitor does a blocking property-get on construction, after which it assumes that any further change calls can be caught by restrictContext

Reply:
Sure, but couldn't you accomplish the same thing without the monitor by storing the "c++" side version of the bool directly, instead of using the monitor?  You'd still do `_isRestricted =isRestrictedGetProperty();` in the initializer, and you'd still catch the changes here.  The monitor just seems like added complexity (unless the signalling is actually necessary to solve the bug, but from your explanation, it seems like that's not the case).

Reply:
If we could guarantee that we could locate all the ContextAwareProfile  objects that were in existence at the time that the value changed (or guaranteed that there was only one value for that specific QQmlContext) then I would agree with you, otherwise this is a way to distribute the changes to all related objects.

Reply:
Could we keep the static map, but make the values the ContextAwareProfiles?

Reply:
Wait, also, wasn’t the original code one value per thing?  How did we keep them synchronized before?

Reply:
I think prior to this patch the only official value was in QML, each ContextAwareProfile would be querying the QML profile with a one-second cache time.  I'm guessing the cache was there for performance reasons, but that doesn't really help if the cache runs out at the wrong time.

I'm thinking we can't really change the static map to ContextAwareProfiles unless we make it something like map<QQmlContext,list<ContextAwareProfile>>.  We then also start having a potential lock delay as construction/destruction of ContextAwareProfile would require mutexed access to a static object.  Again it depends on how many ContextAwareProfile objects there are vs QQmlContext objects they are associated with.

Reply:
I still don’t really understand the need for the map.  Before, each ContentAwareProfile queried their context for the value.  Can’t we keep that, but only query the values from QML at initialization, and then keep track of the changes in a local member variable?  I see what you’re saying about it not working if there’s multiple Profiles created from one Context, but that case wasn’t handled by the old code, which makes me think it can’t really happen, or isn’t important if it does

Reply:
The issue is in the "keep track of the changes in a local member variable". ContextAwareProfile::restrictContext() is a static method and only specifies QQmlContext*, so unless there's a way of storing something new in the QQmlContext (which seems like a generic Qt interface), we're gonna need a static map in any case if we're ever expecting to reach ContextAwareProfile.  The only real question is what it should map to, i created a new class to store the value for the QQmlContext that permits a Qt subscription.

Reply:
OH fuck me sorry I didn’t see that it was static, you’re totally right.  can we try something like...std::unordered_map<QQmlContext*, std::unordered_set<ContextAwareProfile*>> (or the Qt equivalents, but at least a QHash instead of a QMap)

Reply:
Thank you for going down this rabbit hole with me, this really did need a second pair of eyes.

The monitor object in this file is serving a purpose similar to the std::unordered_map you are suggesting.  Pros and cons of each approach:

monitor class:
- uses the Qt signal/slot approach as a pub/sub to handle the change broadcasts for us
- adds another class to the file that could be considered additional complexity
- would need to hold a static mutex whenever a QQmlContext is seen for the first time, or the last ContextAwareProfile using a specific QQmlContext is destructed

std::unordered_map:
 - would be responsible for iterating over all the entries in the map whenever a change needs to be broadcast
 - would need to hold a static mutex whenever a new ContextAwareProfile is constructed or destructed, or a broadcast is in progress (potential lock contention)

Reply:
Can we try the map approach?  I think it will make this code a lot more readable.  I’m not super concerned about the number of these things that we make (looking at the rest of the code, it doesn’t seem like a problem) or the locking, since if there is locking around just the map access, it should be pretty quick.